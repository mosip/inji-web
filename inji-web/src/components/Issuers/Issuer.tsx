import React from "react";
import {getIssuerDisplayObjectForCurrentLanguage} from "../../utils/i18n";
import {useNavigate} from "react-router-dom";
import {ItemBox} from "../Common/ItemBox";
import {IssuerProps} from "../../types/components";
import {useSelector} from "react-redux";
import {RootState} from "../../types/redux";
import { IssuerWellknownDisplayArrayObject } from "../../types/data";

export const Issuer: React.FC<IssuerProps> = ({issuer, index}) => {
    const language = useSelector((state: RootState) => state.common.language);
    const issuerDisplayObject: IssuerWellknownDisplayArrayObject = getIssuerDisplayObjectForCurrentLanguage(issuer.display, language);
    const navigate = useNavigate();
    return <React.Fragment>
        <ItemBox index={index}
                 url={issuerDisplayObject?.logo?.url}
                 title={issuerDisplayObject?.name}
                 description={issuerDisplayObject?.description}
                 onClick={() => navigate(`/issuers/${issuer.issuer_id}`)}/>
    </React.Fragment>
}

