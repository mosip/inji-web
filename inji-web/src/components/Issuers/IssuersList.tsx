import React from "react";
import {Issuer} from "./Issuer";
import {useSelector} from "react-redux";
import {IssuerObject} from "../../types/data";
import {RootState} from "../../types/redux";
import {useTranslation} from "react-i18next";
import {EmptyListContainer} from "../Common/EmptyListContainer";
import {RequestStatus} from "../../hooks/useFetch";
import {SpinningLoader} from "../Common/SpinningLoader";
import {IssuersListProps} from "../../types/components";
import {HeaderTile} from "../Common/HeaderTile";

export const IssuersList: React.FC<IssuersListProps> = ({state}) => {
    const issuers = useSelector((state: RootState) => state.issuers);
    const {t} = useTranslation("HomePage");

    if (state === RequestStatus.LOADING) {
        return <SpinningLoader/>
    }

    if(state === RequestStatus.ERROR || !issuers?.filtered_issuers || (issuers?.filtered_issuers && issuers?.filtered_issuers?.length === 0)) {
        return <div data-testid="Issuers-List-Container">
            <HeaderTile content={t("containerHeading")}/>
            <EmptyListContainer content={t("emptyContainerContent")}/>
        </div>
    }

    return <React.Fragment>
        <div data-testid="Issuers-List-Container" className={"mb-20"}>
            <HeaderTile content={t("containerHeading")}/>
            <div className="grid grid-cols-1 md:grid-cols-3 sm:grid-cols-2 gap-5">
                {issuers.filtered_issuers.map((issuer: IssuerObject, index: number) =>
                    <Issuer issuer={issuer} key={index} index={index}/>)}
            </div>
        </div>
    </React.Fragment>
}

