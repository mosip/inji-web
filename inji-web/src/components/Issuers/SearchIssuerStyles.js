export const SearchIssuerStyles = {
	mainContainer: "flex flex-col justify-center items-center w-full",
	searchContainer: "w-full sm:w-4/4 md:w-3/4 flex justify-start items-center bg-iw-background shadow-iw rounded-lg",
	searchIcon: "m-2 sm:m-3 md:m-4 lg:m-5",
	searchInput: "py-2 sm:py-3 md:py-4 lg:py-6 w-3/4 sm:w-11/12 text-iw-searchTitle focus:outline-none overflow-ellipsis ml-0 mr-2 sm:mr-3 md:mr-4 lg:mr-10",
	clearSearchIcon: "m-2 sm:m-3 md:m-4 lg:m-5 cursor-pointer",
	helperAndErrorTextContainer: "text-left w-full sm:w-4/4 md:w-3/4 mt-3",
  regexError: "text-iw-red text-sm-plus",
  infoMsg: "text-iw-closeIcon text-sm-plus"
};