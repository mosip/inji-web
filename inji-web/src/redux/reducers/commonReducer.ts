import {storage} from "../../utils/storage";
import {CommonReducerActionType} from "../../types/redux";

const initialState = {
    language: storage.getItem(storage.SELECTED_LANGUAGE) ? storage.getItem(storage.SELECTED_LANGUAGE) : 'en',
}

const CommonReducerAction: CommonReducerActionType = {
    STORE_LANGUAGE: 'STORE_LANGUAGE',
}

export const commonReducer = (state = initialState, actions: any) => {
    switch (actions.type) {
        case CommonReducerAction.STORE_LANGUAGE: {
            return {
                ...state,
                language: actions.language
            }
        }
        default :
            return state;
    }
}

export const storeLanguage = (language: string) => {
    return {
        type: CommonReducerAction.STORE_LANGUAGE,
        language: language
    }
}


