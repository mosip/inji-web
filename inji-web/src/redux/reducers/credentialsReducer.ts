import {CredentialsReducerActionType} from "../../types/redux";

const initialState = {
    credentials: []
}

const CredentialsReducerAction: CredentialsReducerActionType = {
    STORE_CREDENTIAL: 'STORE_CREDENTIAL'
}

export const credentialsReducer = (state = initialState, action: any) => {
    switch (action.type) {
        case CredentialsReducerAction.STORE_CREDENTIAL :
            return {
                ...state,
                credentials: action.credentials
            }
        default :
            return state;
    }
}

export const storeCredentials = (credentials: any) => {
    return {
        type: CredentialsReducerAction.STORE_CREDENTIAL,
        credentials: credentials
    }
}
