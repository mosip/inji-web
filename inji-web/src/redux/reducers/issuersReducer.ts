import {IssuersReducerActionType} from "../../types/redux";

const initialState = {
    issuers: [],
    selected_issuer: {}
}

const IssuersReducerAction: IssuersReducerActionType = {
    STORE_ISSUERS: "STORE_ISSUERS",
    STORE_SELECTED_ISSUER: "STORE_SELECTED_ISSUER"
}

export const issuersReducer = (state = initialState, action: any) => {
    switch (action.type) {
        case IssuersReducerAction.STORE_ISSUERS :
            return {
                ...state,
                issuers: action.issuers
            }
        case IssuersReducerAction.STORE_SELECTED_ISSUER :
            return {
                ...state,
                selected_issuer: action.issuers
            }
        default:
            return state
    }
}

export const storeIssuers = (issuers: any) => {
    return {
        type: "STORE_ISSUERS",
        issuers: issuers
    }
}

export const storeSelectedIssuer = (issuer: any) => {
    return {
        type: "STORE_SELECTED_ISSUER",
        issuers: issuer
    }
}

