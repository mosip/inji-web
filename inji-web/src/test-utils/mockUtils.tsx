import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

