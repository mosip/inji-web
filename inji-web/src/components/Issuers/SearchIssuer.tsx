import React, {useState} from "react";
import {useTranslation} from "react-i18next";
import {FaSearch} from 'react-icons/fa';
import {useDispatch, useSelector} from "react-redux";
import {storeFilteredIssuers} from "../../redux/reducers/issuersReducer";
import {IssuerObject} from "../../types/data";
import {IoCloseCircleSharp} from "react-icons/io5";
import {RootState} from "../../types/redux";
import {getIssuerDisplayObjectForCurrentLanguage} from "../../utils/i18n";
import { SearchIssuerStyles } from "./SearchIssuerStyles";

export const SearchIssuer: React.FC = () => {

    const {t} = useTranslation("IssuersPage");
    const dispatch = useDispatch();
    const [searchText, setSearchText] = useState("");
    const issuers = useSelector((state:RootState) => state.issuers.issuers);
    const language = useSelector((state:RootState) => state.common.language);
    const filterIssuers = async (searchText: string) => {
        setSearchText(searchText);
        setIsSearchValid(issuerRegex.test(searchText));
        const filteredIssuers = issuers.filter( (issuer:IssuerObject) => {
            const displayObject = getIssuerDisplayObjectForCurrentLanguage(issuer.display, language);
            return (displayObject?.name?.toLowerCase().indexOf(searchText.toLowerCase()) !== -1 && issuer.protocol !== 'OTP')
        })
        dispatch(storeFilteredIssuers(filteredIssuers));
    }
    const issuerRegex = /^(?!_)(?!.*__)(?!\s{1,}$)[a-zA-Z0-9\s\-_()]*$/;
    const [isSearchValid, setIsSearchValid] = useState(true);

    return (
        <div className={SearchIssuerStyles.mainContainer}>
            <div
                data-testid="Search-Issuer-Container"
                className={SearchIssuerStyles.searchContainer}
            >
                <FaSearch
                    data-testid="Search-Issuer-Search-Icon"
                    color={"var(--iw-color-searchIcon)"}
                    className={SearchIssuerStyles.searchIcon}
                    size={20}
                />
                <input
                    data-testid="Search-Issuer-Input"
                    type="text"
                    value={searchText}
                    placeholder={t("Intro.searchText")}
                    onChange={(event) => filterIssuers(event.target.value)}
                    className={SearchIssuerStyles.searchInput}
                />
                {searchText.length > 0 && (
                    <IoCloseCircleSharp
                        data-testid="Search-Issuer-Clear-Icon"
                        onClick={() => filterIssuers("")}
                        className={SearchIssuerStyles.clearSearchIcon}
                        color={"var(--iw-color-closeIcon)"}
                        size={20}
                    />
                )}
            </div>
            <div className={SearchIssuerStyles.helperAndErrorTextContainer} data-testid="Search-Issuer-HelperAndError-Text">
            {!isSearchValid && (
                <p className={SearchIssuerStyles.regexError}>
                    { t("searchIssuer.helpText.error") }
                </p>)
            }
            {isSearchValid && (
                <p className={SearchIssuerStyles.infoMsg}>
                    { t("searchIssuer.helpText.info") }
                </p>)
            }
            </div>
        </div>
    );
}

