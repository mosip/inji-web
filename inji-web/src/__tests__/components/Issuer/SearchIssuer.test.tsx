import React from 'react';
import {fireEvent, screen} from '@testing-library/react';
import {SearchIssuer} from '../../../components/Issuers/SearchIssuer';
import {reduxStore} from '../../../redux/reduxStore';
import {storeFilteredIssuers} from '../../../redux/reducers/issuersReducer';
import {mockUseTranslation, renderWithProvider} from '../../../test-utils/mockUtils';

mockUseTranslation();

describe("Testing the layout of SearchIssuer", () => {
    test('Check if the layout is matching with the snapshots', () => {
        const {asFragment} = renderWithProvider(
            <SearchIssuer/>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Testing the Functionality of SearchIssuer', () => {
    beforeEach(() => {
        reduxStore.dispatch = jest.fn();
    });

    test('Check if it filters issuers based on search text', () => {
        renderWithProvider(
            <SearchIssuer/>
        );

        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, {target: {value: 'Issuer 1'}});

        expect(reduxStore.dispatch).toHaveBeenCalledWith(storeFilteredIssuers([]));
    });

    test('Check if it clears search text when clear icon is clicked', () => {
        renderWithProvider(
            <SearchIssuer/>
        );

        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, {target: {value: 'Issuer 1'}});
        expect(input).toHaveValue('Issuer 1');

        const clearIcon = screen.getByTestId('Search-Issuer-Clear-Icon');
        fireEvent.click(clearIcon);
        expect(input).toHaveValue('');
        expect(reduxStore.dispatch).toHaveBeenCalledWith(storeFilteredIssuers([]));
    });
    afterEach(() => {
        jest.clearAllMocks();
    })
});

describe('Testing the Regex Validation and Helper Text of SearchIssuer', () => {
    beforeEach(() => {
        reduxStore.dispatch = jest.fn();
    });

    test('Shows default helper text when input is empty', () => {
        renderWithProvider(<SearchIssuer />);

        expect(screen.getByText("Search by issuer name using letters, numbers, hyphens (-), underscores (_) or brackets ( ) only.")).toBeInTheDocument;
    });

    test('Shows default helper text when input is valid', () => {
        renderWithProvider(<SearchIssuer />);
        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, { target: { value: 'AaZz09 _-()' } });

        expect(screen.queryByText("Search by issuer name using letters, numbers, hyphens (-), underscores (_) or brackets ( ) only.")).toBeInTheDocument();
        expect(screen.queryByText("Please enter a valid issuer name. Only letters, numbers, spaces, hyphens (-), underscores (_), and brackets ( ) are allowed. Special characters are not permitted.")).not.toBeInTheDocument();
    });

    test('Shows error helper text when input is invalid', () => {
        renderWithProvider(<SearchIssuer />);
        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, { target: { value: 'Issuer@#' } });

        expect(screen.queryByText("Please enter a valid issuer name. Only letters, numbers, spaces, hyphens (-), underscores (_), and brackets ( ) are allowed. Special characters are not permitted.")).toBeInTheDocument();
    });

    test('Shows error when input is only spaces', () => {
        renderWithProvider(<SearchIssuer />);
        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, { target: { value: '     ' } });

        expect(screen.queryByText(
            "Please enter a valid issuer name. Only letters, numbers, spaces, hyphens (-), underscores (_), and brackets ( ) are allowed. Special characters are not permitted."
        )).toBeInTheDocument();
    });

    test('Shows error when input has two underscores', () => {
        renderWithProvider(<SearchIssuer />);
        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, { target: { value: 'Issuer__Name' } });

        expect(screen.queryByText(
            "Please enter a valid issuer name. Only letters, numbers, spaces, hyphens (-), underscores (_), and brackets ( ) are allowed. Special characters are not permitted."
        )).toBeInTheDocument();
    });

    test('Shows error when input starts with underscore (_Issuer)', () => {
        renderWithProvider(<SearchIssuer />);
        const input = screen.getByTestId('Search-Issuer-Input');
        fireEvent.change(input, { target: { value: '_Issuer' } });

        expect(screen.queryByText(
            "Please enter a valid issuer name. Only letters, numbers, spaces, hyphens (-), underscores (_), and brackets ( ) are allowed. Special characters are not permitted."
        )).toBeInTheDocument();
    });
});