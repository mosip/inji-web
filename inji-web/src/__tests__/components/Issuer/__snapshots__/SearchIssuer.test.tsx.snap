// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the layout of SearchIssuer Check if the layout is matching with the snapshots 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center w-full"
  >
    <div
      class="w-full sm:w-4/4 md:w-3/4 flex justify-start items-center bg-iw-background shadow-iw rounded-lg"
      data-testid="Search-Issuer-Container"
    >
      <svg
        class="m-2 sm:m-3 md:m-4 lg:m-5"
        color="var(--iw-color-searchIcon)"
        data-testid="Search-Issuer-Search-Icon"
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        class="py-2 sm:py-3 md:py-4 lg:py-6 w-3/4 sm:w-11/12 text-iw-searchTitle focus:outline-none overflow-ellipsis ml-0 mr-2 sm:mr-3 md:mr-4 lg:mr-10"
        data-testid="Search-Issuer-Input"
        placeholder="Search Issuers by name"
        type="text"
        value=""
      />
    </div>
    <div
      class="text-left w-full sm:w-4/4 md:w-3/4 mt-3"
      data-testid="Search-Issuer-HelperAndError-Text"
    >
      <p
        class="text-iw-closeIcon text-sm-plus"
      >
        Search by issuer name using letters, numbers, hyphens (-), underscores (_) or brackets ( ) only.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
